### PR TITLE
[WIP] Move most manga endpoints to API 3

### PR DESCRIPTION
**Pull request by null-dev**
So I've moved most of the catalogue, cover images, library and manga info to the v3 API. That's about all that I'm willing to move for now (because I don't want to move everything over until v3 is completely implemented). I also won't be implementing any new features in this PR (e.g. bookmarking chapters) and I'll probably just leave those features to you for the time being.

Also just made a slight change so chapters and manga info are fetched in parallel on the manga-info page.

TODO:
- [x] Fix library download filter (broken during the move)
- [x] Fix library sort by total chapters (broken during the move)
- [x] Fix library sort by source name (broken during the move)
- [x] Fix library sort by last read
- [x] Fix manga status detail display in manga info (broken during the move)
- [x] Fix manga source name detail display in manga info (broken during the move)
- [x] ~~Fix manga categories detail display in manga info (broken during the move)~~ **Feature removed**
- [x] ~~Fix `Open source website` in manga info (broken during the move)~~ **Fixed on server**

Requires the latest version of the server binary to work properly.

### CODE REVIEW

File: src/components/library/LibrarySort.js [hunk 1]
```diff
@@ -12,9 +12,10 @@ import { withStyles } from "@material-ui/core/styles";
 const sorts = [
   { flagState: "ALPHA", description: "Alphabetically" },
   // "LAST_READ" and "LAST_UPDATED" not yet implemented
-  { flagState: "UNREAD", description: "Unread" },
-  { flagState: "TOTAL", description: "Total chapters" },
-  { flagState: "SOURCE", description: "Source" }
+  { flagState: "LAST_UPDATED", description: "Last updated" },
+  { flagState: "UNREAD", description: "Unread" }
+  // { flagState: "TOTAL", description: "Total chapters" },
```
Comment (Ligerx) [inline src/components/library/LibrarySort.js]: Why have these sorts been commented out?

File: src/pages/MangaInfo.js [hunk 2]
```diff
@@ -1,48 +1,44 @@
...
+      }
+      return null;
+    });
+
+    fetchMangaInfo().then(() => {
+      const { mangaInfo } = this.props;
+      // Update manga manually if the server has not fetched the data for
+      //   this manga yet
+      if (mangaInfo && !mangaInfo.initialized) {
+        updateMangaInfo();
```
Comment (Ligerx) [inline src/pages/MangaInfo.js]: Are you sure this will work without the setTimeout?
I remember when I was developing this part that the delay was necessary for the info to fetch.

File: src/redux-ducks/library.js [hunk 3]
```diff
@@ -2,7 +2,7 @@
 import { Server } from "api";
 import type { LibraryFlagsType, LibraryFlagsPossibleValueTypes } from "types";
 import { ADD_MANGA } from "./mangaInfos";
-import { handleHTMLError, transformToMangaIdsArray } from "./utils";
+import { handleHTMLError, transformToMangaIdsArray } from "./utils"; import type {LibraryManga} from "@tachiweb/api-client";
```
Comment (Ligerx) [inline src/redux-ducks/library.js]: add new line

**Comment by Ligerx** [comment]
Leaving a comment here because I left some comments and they aren't showing up in this view.

File: src/pages/MangaInfo.js (hunk 2 above)
Comment (null-dev) [reply src/pages/MangaInfo.js]: I'm not sure why the extra time was needed in the first place. The endpoint being called by `fetchMangaInfo` is only a couple lines of code: https://github.com/TachiWeb/TachiWeb-Server/blob/4a6b3335191b08b6ad8d2db4ce6de88ce9c3179c/TachiServer/src/main/java/xyz/nulldev/ts/api/http/manga/MangaRoute.kt#L36-L41 . All it does is get the manga from the database and serialize it into a JSON. It doesn't do any kind of scraping...

The v3 version of this endpoint that I've moved to does the exact same thing.

Either way, I didn't notice any problems after I made this change so it could have been a bug that was specific to the v1/v2 endpoints...

File: src/redux-ducks/library.js (hunk 3 above)
Comment (null-dev) [reply src/redux-ducks/library.js]: Nice catch! I didn't look at all the eslint errors before I commited so I probably missed this.

File: src/components/library/LibrarySort.js (hunk 1 above)
Comment (null-dev) [reply src/components/library/LibrarySort.js]: The total chapters are no longer supplied as a field in the manga object for performance reasons. The tachiyomi app works around this by only loading the total chapters when the `TOTAL` sort option is selected. I'm currently thinking about how I can architect this so we can still support the `TOTAL` sort option while keeping performance good for large libraries.

The `SOURCE` sort option is supposed to sort manga by their source names. But with the v3 API, the source name is no longer provided with the manga object, instead the source id provided. So I have to fetch the sources before we have enough information to use this sort.

File: src/pages/MangaInfo.js (hunk 2 above)
Comment (Ligerx) [reply src/pages/MangaInfo.js]: Not sure if this helps, I believe I faced this problem when I started searching something new in the catalogue and tried to view a manga really fast. My guess is that somewhere in that chain of events the server is spending time scraping the source website, and because that takes time, it's possible for `fetchMangaInfo` to return almost no information

Comment (null-dev) [reply src/pages/MangaInfo.js]: Hmm, I've tried to replicate this but I still can't. I actually did rework most of the async/threading portions of the code from API v1/v2 to v3 so that's why I said maybe it was a bug specific to the v1/v2 endpoints.

I think we should leave the delay out for now and if it comes up again, we'll throw it back in.

File: src/pages/Catalogue.js [hunk 4]
```diff
@@ -17,9 +17,6 @@ import { Helmet } from 'react-helmet';
 
 // TODO: keep previous scroll position when going back from MangaInfo -> Catalogue
 
-// FIXME: If you type something into the search bar,
```
Comment (null-dev) [inline src/pages/Catalogue.js]: Fixed this on the server side

File: src/redux-ducks/mangaInfos.js [hunk 5]
```diff
@@ -141,16 +137,15 @@ export function toggleFavorite(mangaId: number, isCurrentlyFavorite: boolean) {
       meta: { mangaId, isCurrentlyFavorite }
     });
 
-    return fetch(Server.toggleFavorite(mangaId, isCurrentlyFavorite))
-      .then(handleHTMLError)
+    // TODO Remove toString when https://github.com/OpenAPITools/openapi-generator/pull/2499 is merged
```
Comment (null-dev) [inline src/redux-ducks/mangaInfos.js]: This is now merged, but not in a release yet so I'm going to keep this comment.

File: src/components/catalogue/CatalogueHeader.js [hunk 6]
```diff
@@ -53,7 +53,9 @@ const CatalogueHeader = ({
             >
               {sources.map((source, index) => (
                 <MenuItem value={index} key={source.id}>
-                  {source.name}
+                  {(source.lang != null
```
Comment (null-dev) [inline src/components/catalogue/CatalogueHeader.js]: Switching to the v3 sources endpoint causes all the sources to be enabled and the catalogue was confusing since it looked like there were several copies of each source even though it was just the same source in different languages. I've added this language indicator to workaround this problem for now but it would be better if we had a screen to enable/disable sources.

**Comment by null-dev** [comment]
This PR is pretty much done so if it looks good to you @Ligerx feel free to squash + merge!

File: src/redux-ducks/library.js [hunk 7]
```diff
@@ -54,7 +64,13 @@ const defaultState: State = {
   mangaIds: [], // array of mangaIds that point at data loaded in mangaInfos reducer
   reloadLibrary: true, // Library should be loaded once on first visit
   unread: {}, // { mangaId: int }
+  downloaded: {},
+  totalChaptersSortIndexes: {},
+  lastReadSortIndexes: {},
   reloadUnread: true, // should refresh unread for library if something new is added
+  reloadDownloaded: true,
```
Comment (null-dev) [inline src/redux-ducks/library.js]: I will be splitting up the `library` endpoint to make these individually reloadable in the future.

**Comment by Ligerx** [comment]
_this was a comment I wrote a couple weeks ago but forgot to submit_
I noticed something unusual in redux when looking at the library page (empty library).

There's a dispatch for both `"library/FETCH_UNREAD_SUCCESS"` and `"library/FETCH_UNREAD_CACHE"` but not `"library/FETCH_UNREAD_REQUEST"`

image

**Comment by Ligerx** [comment]
I also noticed that since you've moved some Types over to `"@tachiweb/api-client"`, Flow has been resolving those types to `any`. And since Flow is only in-editor error checking, types basically become useless.

This actually seems like kind of a big problem. Any idea how to fix this?

File: src/redux-ducks/extensions.js [hunk 8]
```diff
@@ -102,6 +103,7 @@ export function installExtension(packageName: string) {
 
       dispatch({ type: INSTALL_SUCCESS, extension });
       dispatch({ type: RESET_CATALOGUE_STATE });
+      await fetchSources();
```
[the remark this answers is not in the document]
Comment (Ligerx) [reply src/redux-ducks/extensions.js]: I think the issue is that it needs to be wrapped in a `dispatch(...)`?

[the remark this answers is not in the document]
Comment (Ligerx) [reply src/redux-ducks/extensions.js]: This doesn't resolve the caching issue though

[the remark this answers is not in the document]
Comment (null-dev) [reply src/redux-ducks/extensions.js]: We fetch all the sources again because when we install an extension there is a new sources and we don't know what those sources are. I manually remove sources on uninstall because we know what sources are associated with each installed extension so we know those sources will be gone once the extension is uninstalled.

File: src/components/catalogue/CatalogueHeader.js [hunk 9]
```diff
@@ -53,7 +53,9 @@ const CatalogueHeader = ({
...
+                    ? `(${source.lang.toUpperCase()}) `
```
[the remark this answers is not in the document]
Comment (null-dev) [reply src/components/catalogue/CatalogueHeader.js]: So the sources would display as: `MangaDexEN` instead of `MangaDex (EN)`? Also, the single quotation marks are a typo right?

File: src/components/mangaInfo/MangaInfoDetails.js [hunk 10]
```diff
@@ -66,8 +79,8 @@ MangaInfoDetails.defaultProps = {
 };
 
 // Helper functions
-function detailsElements(mangaInfo: MangaType): React.Node {
-  const fieldNames = ['status', 'source', 'author', 'genres', 'categories'];
+function detailsElements(mangaInfo: Manga): React.Node {
+  const fieldNames = ["author", "artist", "genre"];
```
[the remark this answers is not in the document]
Comment (null-dev) [reply src/components/mangaInfo/MangaInfoDetails.js]: I removed categories because in order to get that information with the new v3 API, we have to make another API call and since Tachiyomi doesn't display the categories on the manga info screen anyways, I didn't feel like that was needed in TachiWeb.

The categories information use to come with the manga-info API call but is no longer present in the API v3 version of that call. It was removed because it actually involves making another DB query (separate from the one that gets the manga info) which although fairly cheap in other applications, is expensive in TachiWeb due to Tachiyomi's architecture.

Basically, all the v3 endpoints are designed in such a way that `one API call = about one DB query`.

File: src/pages/Catalogue.js [hunk 11]
```diff
@@ -17,9 +17,6 @@ import { Helmet } from 'react-helmet';
```
[the remark this answers is not in the document]
Comment (null-dev) [reply src/pages/Catalogue.js]: If I make edits to an existing import, usually I will keep it as single quotes but since the project settings are double quotes if I make any major edits to the imports I will change them to double quotes. I didn't make any edits to those lines though so they are kept as single quotes.

File: src/redux-ducks/mangaInfos.js [hunk 12]
```diff
@@ -40,22 +39,23 @@ const SET_FLAG_NO_CHANGE = "mangaInfos/SET_FLAG_NO_CHANGE";
...
+  action: Object = {}
+) {
   switch (action.type) {
     case ADD_MANGA:
       return { ...state, ...mangaArrayToObject(action.newManga) };
 
     case FETCH_MANGA_CACHE:
       return state;
 
+    case UPDATE_MANGA_SUCCESS:
```
[the remark this answers is not in the document]
Comment (null-dev) [reply src/redux-ducks/mangaInfos.js]: This case will actually fallthough. A `switch` statement will continue to execute lines of code underneath a matching `case` statement until it reaches either a `break` or a `return` statement.

File: src/components/mangaInfo/MangaInfoDetails.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (null-dev) [reply src/components/mangaInfo/MangaInfoDetails.js]: That said, yeah, I agree it should be refactored, I'll move the detail section into it's own component.

File: src/pages/Catalogue.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (Ligerx) [reply src/pages/Catalogue.js]: Are the automatic linting settings not working for you?

**Comment by Ligerx** [comment]
I left this comment previously

> I also noticed that since you've moved some Types over to "@tachiweb/api-client", Flow has been resolving those types to any. And since Flow is only in-editor type checking, types basically become useless.
> 
> This actually seems like kind of a big problem. Any idea how to fix this?

But I just noticed that flow is not working correctly on my local machine, so maybe this is just a personal problem

**Comment by Ligerx** [comment]
Confirmed, Flow is just broken. Downgrading works. https://github.com/Ligerx/TachiWeb-React/issues/66

Any comments I left about not being able to see your types can be ignored

**Comment by Ligerx** [comment]
I've been rewriting a ton of the codebase. There's gonna be a ton of merge conflicts so I'm going to start manually merging this PR.
